From edbc2096e79fd51b36fb5d23c909d9ca105e0889 Mon Sep 17 00:00:00 2001
From: Ken Zhu <quic_guigenz@quicinc.com>
Date: Tue, 20 Jun 2023 13:11:11 -0700
Subject: [PATCH] linux: fix ipv6 address error for unusual icmp message in
 ipip6 tunnel

the ipv6 destination address is wrong for unusual icmp message, such as
unreachable, redirect etc. type message.
Rather than echo/reply icmp packet, unusual icmp message needs to generate
the ipv6 address from the original ipv4 header inside its packet.

Change-Id: Iaa69e7907a273c0b4d65101d834ce1949eabe3f8
Signed-off-by: Ken Zhu <guigenz@codeaurora.org>
Signed-off-by: John Sanli <quic_jsanli@quicinc.com>
---
 net/ipv6/ip6_tunnel.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/ipv6/ip6_tunnel.c b/net/ipv6/ip6_tunnel.c
index 8f3ee2ae7e19..12b9b3e7cf6a 100644
--- a/net/ipv6/ip6_tunnel.c
+++ b/net/ipv6/ip6_tunnel.c
@@ -883,7 +883,7 @@ static void ip4ip6_fmr_calc(struct in6_addr *dest,
 		int frombyte = fmr->ip6_prefix_len / 8;
 		int fromrem = fmr->ip6_prefix_len % 8;
 		int bytes = sizeof(struct in6_addr) - frombyte;
-		const u32 *addr = (use_dest_addr) ? &iph->daddr : &iph->saddr;
+		const u32 *addr = (use_dest_addr) ? &dsth->daddr : &dsth->saddr;
 		u64 eabits = ((u64)ntohl(*addr)) << (32 + fmr->ip4_prefix_len);
 		u64 t = 0;
 
-- 
2.34.1

